### PR TITLE
feat(quota,api): APIs for quotas

### PR DESCRIPTION
**Pull request by heww**
Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: src/common/dao/project/projectmember.go [hunk 1]
```diff
@@ -70,6 +70,59 @@ func GetProjectMember(queryMember models.Member) ([]*models.Member, error) {
...
+SELECT
+  COUNT(1)
+FROM
+  (
+    SELECT
+      pm.id AS id,
+      pm.project_id AS project_id,
+      r.role_id AS role
+    FROM
+      user_group ug
```
Comment (ywk253100) [inline src/common/dao/project/projectmember.go]: The group can be a LDAP group, and you cannot list users under the group. So should we show the users under the group or just show the count of users + groups?

Comment (heww) [reply src/common/dao/project/projectmember.go]: Fixed

File: src/common/dao/project/projectmember.go [hunk 2]
```diff
@@ -70,6 +70,27 @@ func GetProjectMember(queryMember models.Member) ([]*models.Member, error) {
...
+
+	queryParam := []interface{}{projectID}
+
+	if len(roles) > 0 {
+		sql += " AND role = ?"
+		queryParam = append(queryParam, roles[0])
+	}
+
+	var count int64
+	err := dao.GetOrmer().Raw(sql, queryParam).QueryRow(&count)
```
Comment (wy65701436) [inline src/common/dao/project/projectmember.go]: use .count() method?

File: src/common/dao/quota.go [hunk 3]
```diff
@@ -84,58 +132,99 @@ func ListQuotas(query ...*models.QuotaQuery) ([]*models.Quota, error) {
...
 		return nil, err
 	}
 
+	for _, quota := range quotas {
+		d, ok := driver.Get(quota.Reference)
+		if !ok {
+			continue
+		}
+
+		ref, err := d.Load(quota.ReferenceID)
```
Comment (wy65701436) [inline src/common/dao/quota.go]: add a warning log to index the reference is missing for this referenceID?

File: src/common/dao/quota.go [hunk 4]
```diff
@@ -84,58 +132,99 @@ func ListQuotas(query ...*models.QuotaQuery) ([]*models.Quota, error) {
...
 	return quotas, nil
 }
 
+// GetTotalOfQuotas returns total of quotas
+func GetTotalOfQuotas(query ...*models.QuotaQuery) (int64, error) {
+	condition, params := quotaQueryConditions(query...)
+	sql := fmt.Sprintf("SELECT COUNT(1) FROM quota AS a JOIN quota_usage AS b ON a.id = b.id %s", condition)
+
+	var count int64
+	if err := GetOrmer().Raw(sql, params).QueryRow(&count); err != nil {
```
Comment (wy65701436) [inline src/common/dao/quota.go]: use .count() instead?

File: src/core/api/quota.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+func (qa *QuotaAPI) Put() {
+	var req *models.QuotaUpdateRequest
+	if err := qa.DecodeJSONReq(&req); err != nil {
+		qa.SendBadRequestError(err)
+		return
+	}
+
+	mgr, err := quota.NewManager(qa.quota.Reference, qa.quota.ReferenceID)
+	if err != nil {
+		qa.SendBadRequestError(err)
```
Comment (wy65701436) [inline src/core/api/quota.go]: 500?

File: src/core/api/quota.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+	total, err := dao.GetTotalOfQuotas(query)
+	if err != nil {
+		qa.SendInternalServerError(fmt.Errorf("failed to query database for total of quotas, error: %v", err))
+		return
+	}
+
+	quotas, err := dao.ListQuotas(query)
+	if err != nil {
+		qa.SendInternalServerError(fmt.Errorf("failed to query database for quotas, error: %v", err))
```
Comment (wy65701436) [inline src/core/api/quota.go]: qa.ParseAndHandleError()?

**Pull request merged**